torch/distributed/_fsdp/fully_sharded_data_parallel.py

```diff
@@ -0,0 +1,812 @@
+        previously initialized optimizers.
+    .. warning:
+        Module should be already placed on the destination device or
+        device is set properly using torch.cuda.set_device(device_id).
+        FSDP will get compute device from module first, if module device
+        is CPU, FSDP will then get compute device from current device.
+    Args:
+        module (nn.Module):
+            module to be wrapped with FSDP.
+        process_group (Optional[ProcessGroup]):
+            process group for sharding
+    """
+
+    def __init__(
+        self,
+        module: nn.Module,
+        process_group: Optional[ProcessGroup] = None,
+    ):
+        super().__init__()
+        self.module = module
+        self.process_group = process_group or _get_default_group()
+        self.rank = self.process_group.rank()
+        self.world_size = self.process_group.size()
+        # device for computation, if module is on GPU, use module.device;
+        # if module is on CPU, use current device;
+        self.compute_device = _get_default_cuda_device(module)
+        self.compute_dtype = next(module.parameters()).dtype
+
+        # Free full params and keep shard only after forward
+        self.reshard_after_forward = True
+
+        # setting two factors to avoid underflow and overflow
+        self.gradient_predivide_factor: float = self._get_gradient_predivide_factor(
+            self.world_size
+        )
+        self.gradient_postdivide_factor: float = (
+            self.world_size / self.gradient_predivide_factor
+        )
+
+        self.numel_padded_per_param: List[int] = []
+
+        # Only handle params which are not already sharded. This enables
+        # sharding individual layers of a Module, with an outer wrapper to
+        # shard any leftover parameters.
+        params = []
+        for param_name, param in module.named_parameters():
+            if not hasattr(param, "_is_sharded"):
+                params.append(param)
+        self.params = params
+
+        # Shard module parameters in place
+        self._shard_parameters()
+
+        # Make sure all parameters are sharded.
+        for n, p in self.named_parameters():
+            if not hasattr(p, "_is_sharded"):
+                raise RuntimeError(f"found unsharded parameter: {n} ; {p.size()}")
+        self._reset_lazy_init()
+
+        # Enum to indicate if we're in the forward/backward pass, idle, etc.
+        self.training_state = TrainingState_.IDLE
+
+        # Flag to guard against preparing gradients multiple times per backward pass.
+        self._pre_backward_hook_has_run = False
+
+    # setting two factors 'self.gradient_predivide_factor'
+    # and 'self.gradient_postdivide_factor' to avoid underflow and overflow
+    def _get_gradient_predivide_factor(self, world_size: int) -> float:
+        factor: int = 1
+        while world_size % factor == 0 and world_size / factor > factor:
```

Comment:
I haven't checked in detail - but will the algorithm still be correct for non power of 2 world_size?

Reply:
it works well when world_size is even number, when world_size is odd number, it returns world_size here